### ISSUE TITLE
Dupable wars, episode IV: a new plan

### ISSUE THREAD

**Issue by aspiwack** [tweag/linear-base#343]
## Background

I've been insisting, in the past few years, that the `Dupable` type class ought to:
- Support instance being defined in a `Data.Applicative` style
- Support generating n-tuples for any n.

To this effect, we've been using `dupV :: a %1 -> V n a` as our main type class method.

We realised long ago that the `Data.Applicative` instance for `dupV` is not super efficient, as it allocates intermediate arrays for each use of `(<*>)`. Which I have proposed in the past should be solved by using pull arrays instead (#208) and deducing `dupV` from the pull-array based duplication function.

Another proposal which has been floated is to specialised to `V 2`, which would be sufficient (https://github.com/tweag/linear-base/issues/151#issuecomment-690130887). I find my explanation, in this comment, that `V 2` is insufficient to be lacking: I can defined `dupV` in term of `dup2` (equivalently `dupV2`). I didn't explain why it is not a great idea.

## Movable -> Dupable

A typical instance of `Dupable` will deep-copy the data type. When I defined `dupV @n` in terms of `dup2`, I need to do `n-1` of these copies. This is often ok, but if I have a `Movable` instance, I can do a single copy:

```haskell
dupV x = case move x of
  Ur x' -> pure x')
```

It would be quite a pity to forbid ourselves from using this when we can.

## A streaming hot approach

I was thinking about the choice between non-mutable and mutable array for the implementation of V (#312, #341). And was starting to be a bit dismayed by the seemingly untamable zoology of twisted little data structures all alike. How can we keep the complexity low? Do we need a special, scoped, dup function for the mutable case (https://github.com/tweag/linear-base/pull/341#issuecomment-905344920).

Then I realised that I had made a pretty incorrect assumption there: that the only applicable multi-ary `Data.Applicative` was `V`. This is very not true: there are also (pull-polarity) streams. This would look something like this:

```haskell
data Stream a where
  Stream { state :: x, next :: x %1 -> (a, x), final :: x %1 -> a}

instance Data.Applicative Stream where
  pure a = Stream
    { state = Ur a
    , next = (\(Ur x) -> (x, Ur x))
    , final = unur }
  (fs <*> xs) = Stream
    { state = (state fs, state xs)
    , next = (\(sf, sx) -> case (next sf, next sx) of 
                           ((f, sf'), (x, sx')) -> (f x, (sf', sx')))
    , final = (\(sf, sx) -> final sf (final sx))}
```

(`Stream` is a variant of `AffineStream (Of a) Identity a` from the streaming sublibrary with a single occurrence of `a` (to be able to define instances), `Of` is linear in `a`, and which (crucially) cannot decide to end on its own)

We could define `dupS :: a %1 -> Stream a` as our main method for the `Dupable` type class. Note how `pure` gives us the `move`-based implementation.

We could then derive functions like `dupV` (or the mutable equivalent) as a derived concept. More generally, we can define

```haskell
elim :: KnownNat n => Fun n a b -> Stream a %1 -> b
```

(though the question of doing this efficiently is probably as tricky as in the case of `V.elim`, but it does give us elimination into `V`, and into `tuples` for instance).

This gives us a much more comprehensible `Dupable` type class too! And it becomes the responsibility of the various data types that we provide to provide a good interface to `Dupable`. Much more manageable.

## A moving touch

There is still something unfortunate though (which also holds the current `dupV`-based abstraction): if I use `<*>` on `pure` streams, then I lose the only-one-deep-copy property. Let's look at the instance for lists:

```haskell
instance Dupable a => Dupable [a] where
  dupS [] = pure []
  dupS (a:as) = (:) <&> dupS a <*> dupS as
```

If `a` is movable, then `[a]` is movable too. However, if each of the `a`-s is indeed deep-copied only once, the spine of the list is copied `n` times 🙁 .

I think that the right solution is to do some dynamic dispatch:

```haskell
data BetterStream a where
  Moved :: a -> BetterStream a
  Sequential :: Stream a %1 -> BetterStream a

instance Applicative BetterStream a where
  pure = Moved
  
  (Moved f) <*> (Moved x) = Moved (f x)
  fs <*> xs = (toStream fs) <*> (toStream xs)
  
toStream :: BetterStream a %1 -> Stream a
toStream (Moved x) = pure x
toStream (Sequential xs) = xs
```

We can similary defined `elim` for this type.

## Conclusion (kind of)

I don't feel super great to introduce two extra stream types on top of the existing two. I suppose `Stream` may want to stay internal while `BetterStream` is exposed (with a less silly name). A type of stream defined solely to be the output of `dupS` still does feel a tad unsatisfactory. Still, it's probably the best thing to do. I do wonder though if we don't want `BetterStream` to expose its constructors, so that `V` (and its mutable variant) can take advantage of it for a slightly faster build in the `Moved` case. Or maybe we have a clever `elim'` function with a type like

```haskell
elim' :: (a -> b) -> (Fun n a b) -> BetterStream a %1 -> b
```

This would avoid displaying one additional stream types. I don't quite know yet.

Anyway, this does look like the right way forward, doesn't it?

**Comment by aspiwack** [comment]
I forgot a probably important point: you can define `dupS` in terms of `dup2`:

```haskell
dup2 :: a %1 -> (a, a)

dupS :: a %1 -> BetterStream a
dupS a = Sequential $ BetterStream
  { state = a
  , next = dup2
  , final = id }
```

**Comment by Divesh-Otwani** [comment]
I'm just going to comment that I love this title. I'll give my thoughts on the content a bit later.

**Comment by treeowl** [comment]
I know nothing about linear types, but the deep copying/traversal required all over the place seems rather sad. One option would look something like this:

```haskell
class Dupable a where
  ....
  glump :: Maybe (Dict (Moveable a))

class Consumable a where
   ...
   glamp :: Maybe (Dict (Moveable a))
```

Now we could use something like

```haskell
instance (Dupable a, Dupable b) => Dupable (a, b) where
  dup = case glump @(a, b) of
    Nothing -> ....
    Just Dict -> \ab -> case move ab of Ur ab' -> (ab', ab')
  glump = do
    Dict <- glump @a
    Dict <- glump @b 
    Just Dict
```

I dunno how this all relates to the Applicative stuff you're talking about here.

**Comment by treeowl** [comment]
Another flavor:

```haskell
type family IsMoveable a :: Bool
class GoodBool (a :: Bool)
instance GoodBool 'True
instance GoodBool 'False
class GoodBool (IsMoveable a) => Consumable a where ...
class GoodBool (IsMoveable a) => Dupable a where ...
class (IsMoveable a ~ 'True) => Moveable a where ...
```

This shifts more to the type level, but lots of auxiliary classes show up.

**Comment by aspiwack** [comment]
Thanks for the suggestions.

I'd say that you generally want to avoid deep copies. So when you are using one of these classes you have to be cognisant of their cost on the particular type you are, _e.g._, `dup`-ing (much like `.clone()` in Rust).

That being said they are natural classes and we do want to get them right. I'm not quite sure what problem your suggestions are intended to address, though, could you elaborate?

**Comment by treeowl** [comment]
I don't think my approach above is quite sufficient. I think what I'm really trying to do is work more efficiently with the case where `move` is basically a no-op because the values are pure. In that situation, we really don't want to copy *anything*. Does `Ur` have to be a datatype or can it be a newtype? If the latter, I suspect there may be some options with safe coercions.

**Comment by treeowl** [comment]
By the way: if you're wondering why I didn't pull the "spring" out and use a worker stream type, it's because GHC *still* doesn't know how to unbox existential types. That'd be a potentially significant performance hit.

**Comment by treeowl** [comment]
Now I'm starting to doubt myself. Does my version actually do anything better? I'd better go to bed.

**Comment by aspiwack** [comment]
I think that the problem with your version is that you still get to decompose the non-movable `state` to match the structure of the applicative expression. So I think that you have essentially the same computational behaviour as my version (possibly worse as you decompose some things twice): you decompose the structure until you hit a `Movable` substructure, then you use the `spring` bit.

You may be saved by laziness in some cases (if we use `\~(x,y) -> …` everywhere), but I don't know how to check this.

**Comment by treeowl** [comment]
Sorry, it's not an `Apply`. Needs a custom combiner.

**Comment by treeowl** [comment]
I am becoming convinced that your `BetterStream` is sufficient, at least for the "movable values" situation. I was letting my imagination run a bit wild.

**Comment by treeowl** [comment]
@Divesh-Otwani, well, I was working on the assumption that the streams are only used for duplication. If they're not, that's another story.

**Comment by treeowl** [comment]
If we give up on making the fundamental stream type `Applicative`, we gain the option of revealing the seed.

```haskell
class (Consumable a, s ~ Seed a) => Dupable s a where
  type Seed a
  dupS :: Replicator s a -- for some implementation of Replicator
```

(It's definitely possible to do this with a single-parameter `Dupable` class, but I'm not a fan of the way constraints tend to look in that style.)

Then we can impose the constraints on the seed from outside, which is probably more efficient operationally as well as more flexible:

```haskell
instance Consumable s => Consumable (Replicator s a)
instance Dupable s => Dupable (Replicator s a)
instance Movable s => Movable (Replicator s a)
```

Rather than an `Applicative` instance, we'd get custom functions

```haskell
bleep :: a -> Replicator a a
bloop :: (a %1-> b %1-> c) -> Replicator s a %1-> Replicator t b %1-> Replicator (s, t) c
```

**Comment by treeowl** [comment]
We may actually want

```haskell
class (Consumable a, Dupable s, s ~ Seed a) => Dupable s a where
  type Seed a
  dupS :: Replicator s a
```

because it seems unlikely in practice that the seed type can't be duplicated and consumed generally.

**Comment by Divesh-Otwani** [comment]
The last definition seems elegant to me:
  
```haskell
data Streamish a where
  Streamish :: (s %1 -> a) -> (s %1 -> (s, s)) -> (s %1 -> ()) %-> s %1 -> Streamish a
```

I generally don't like super polymorphic definitions unless we see a strong need. I think it addresses (1) n-generation, (2) an applicative instance, (3) seed-friendly replication if the way you duplicate a moveable value is by making `s` some `Ur a`.

**Comment by treeowl** [comment]
@Divesh-Otwani, I don't see how that gives you friendly stream duplication when `s` is `Movable`. My version is really not "super polymorphic"; I find it more "Haskelly" actually.

**Comment by aspiwack** [comment]
> If we give up on making the fundamental stream type `Applicative`, we gain the option of revealing the seed.

That the streams (well the replicators) be applicative is the point. I'd much rather not have an extra parameter to `Dupable` either.

> @Divesh-Otwani, I don't see how that gives you friendly stream duplication when `s` is `Movable`. My version is really not "super polymorphic"; I find it more "Haskelly" actually.

I think “super polymorphic” was referring to my `forall t. (Applicative t, Traversable t) => …` proposal. This is all with an understanding that a replicator is either a `Streamish a` or an unrestricted `a`, to handled the movable case.

**Comment by aspiwack** [comment]
Oh dear! As I'm posting this, I see that the previous comment was “24 days ago”. Sorry for being so unresponsive.

**Comment by treeowl** [comment]
Okay, I'm convinced to drop the type family thing. How about this:

```haskell
data Duplicator a where
  Duplicator :: Dupable s => s %1-> (s %1-> (a, s)) -> (s  %1-> a) -> Duplicator a
```

That lets us duplicate streams efficiently, terminate early (with `consume`), and otherwise do all the stuff.

My only concern: when `s ~ a`, will having a cyclical reference between the `Dupable` dictionary and its `Duplicator` get in the way of compiler optimizations?

**Comment by aspiwack** [comment]
> My only concern: when `s ~ a`, will having a cyclical reference between the `Dupable` dictionary and its `Duplicator` get in the way of compiler optimizations?

I think that it's probably fair to just store the `s %1-> ()` and `s -> (s, s)` methods.

Or we may want to store an `s %1 -> Duplicator s` method directly.

But trying to make everything mutually recursive with the `Dupable` type seems a bit overkill to me.

Note that when we are using this type, `s` is not known to be movable. So we don't need to be able to optimise for this case.
